From 02044b135079643104df738fdfee49e12287151c Mon Sep 17 00:00:00 2001
From: Guoqing Jiang <Guoqing.Jiang@windriver.com>
Date: Fri, 1 Apr 2011 17:45:45 +0800
Subject: [PATCH 12/21] TI 816x: Add power reset management and clock management support

1. add ti 816x power reset management and clock management support
   common files.
2. create ti 816x specific power reset management header file.

This patch refers to below commits in arago git tree.
git://arago-project.org/git/projects/linux-omap3.git

e1ea60 ti816x: Initial support for TI8168
1558f7 ti816x: Update colck data
20a71d ti816x: Add support for global cold reset through software
c45c59 ti616x: clock: Add local reset control for PCIe and USB

Integrated-by: Guoqing Jiang <Guoqing.Jiang@windriver.com>
---
 arch/arm/mach-omap2/prcm-common.h      |   24 +++++++++++++++++++
 arch/arm/mach-omap2/prcm.c             |    6 +++++
 arch/arm/mach-omap2/prm-regbits-816x.h |   39 ++++++++++++++++++++++++++++++++
 arch/arm/mach-omap2/prm.h              |   10 ++++++++
 4 files changed, 79 insertions(+), 0 deletions(-)
 create mode 100644 arch/arm/mach-omap2/prm-regbits-816x.h

diff --git a/arch/arm/mach-omap2/prcm-common.h b/arch/arm/mach-omap2/prcm-common.h
index 51be679..f402c20 100644
--- a/arch/arm/mach-omap2/prcm-common.h
+++ b/arch/arm/mach-omap2/prcm-common.h
@@ -108,6 +108,30 @@
 #define OMAP4430_PRM_DEVICE_MOD		0x1b00
 #define OMAP4430_PRM_INSTR_MOD		0x1f00
 
+/* TI816X PRCM module offsets */
+/*
+ * FIXME: There are no separate spaced instances of CM & PRM as in OMAPx, but we
+ * could still seggregate the offsets for clarity.
+ */
+#define TI816X_PRM_DEVICE_MOD			0x0000	/* 256B */
+#define TI816X_CM_DEVICE_MOD			0x0100	/* 256B */
+#define TI816X_PRM_OCP_SOCKET_MOD		0x0200	/* 256B */
+#define TI816X_CM_DPLL_MOD			0x0300	/* 256B */
+#define TI816X_CM_ACTIVE_MOD			0x0400	/* 256B */
+#define TI816X_CM_DEFAULT_MOD			0x0500	/* 256B */
+#define TI816X_CM_IVAHD0_MOD			0x0600	/* 256B */
+#define TI816X_CM_IVAHD1_MOD			0x0700	/* 256B */
+#define TI816X_CM_IVAHD2_MOD			0x0800	/* 256B */
+#define TI816X_CM_SGX_MOD			0x0900	/* 256B */
+#define TI816X_PRM_ACTIVE_MOD			0x0a00	/* 256B */
+#define TI816X_PRM_DEFAULT_MOD			0x0b00	/* 256B */
+#define TI816X_PRM_IVAHD0_MOD			0x0c00	/* 256B */
+#define TI816X_PRM_IVAHD1_MOD			0x0d00	/* 256B */
+#define TI816X_PRM_IVAHD2_MOD			0x0e00	/* 256B */
+#define TI816X_PRM_SGX_MOD			0x0f00	/* 256B */
+#define TI816X_CM_ALWON_MOD			0x1400	/* 1KB */
+#define TI816X_PRM_ALWON_MOD			0x1800	/* 1KB */
+
 /* SCRM instances */
 
 #define OMAP4430_SCRM_SCRM_MOD	0x0000
diff --git a/arch/arm/mach-omap2/prcm.c b/arch/arm/mach-omap2/prcm.c
index 07a60f1..2ebb0a7 100644
--- a/arch/arm/mach-omap2/prcm.c
+++ b/arch/arm/mach-omap2/prcm.c
@@ -154,6 +154,8 @@ void omap_prcm_arch_reset(char mode, const char *cmd)
 		omap_writel(l, OMAP343X_SCRATCHPAD + 4);
 	} else if (cpu_is_omap44xx())
 		prcm_offs = OMAP4430_PRM_DEVICE_MOD;
+	else if (cpu_is_ti816x())
+		prcm_offs = 0;
 	else
 		WARN_ON(1);
 
@@ -163,6 +165,10 @@ void omap_prcm_arch_reset(char mode, const char *cmd)
 	if (cpu_is_omap44xx())
 		prm_set_mod_reg_bits(OMAP_RST_DPLL3, prcm_offs,
 						 OMAP4_RM_RSTCTRL);
+
+	if (cpu_is_ti816x())
+		prm_set_mod_reg_bits(TI816X_GLOBAL_RST_COLD, prcm_offs,
+						TI816X_PRM_DEVICE_RSTCTRL);
 }
 
 static inline u32 __omap_prcm_read(void __iomem *base, s16 module, u16 reg)
diff --git a/arch/arm/mach-omap2/prm-regbits-816x.h b/arch/arm/mach-omap2/prm-regbits-816x.h
new file mode 100644
index 0000000..5d98845
--- /dev/null
+++ b/arch/arm/mach-omap2/prm-regbits-816x.h
@@ -0,0 +1,39 @@
+/*
+ * PRM Register bit macros.
+ *
+ * Copyright (C) 2010 Texas Instruments, Inc. - http://www.ti.com/
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation version 2.
+ *
+ * This program is distributed "as is" WITHOUT ANY WARRANTY of any
+ * kind, whether express or implied; without even the implied warranty
+ * of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#ifndef __ARCH_ARM_MACH_OMAP2_PRM_REGBITS_816X_H
+#define __ARCH_ARM_MACH_OMAP2_PRM_REGBITS_816X_H
+
+#include "prm.h"
+
+/*
+ * TODO: At present this file only contains few register values required to
+ * manage local reset for some modules. Later this have to be generated from h/w
+ * database.
+ */
+
+/* Used by RM_DEFAULT_RSTCTRL */
+#define TI816X_USB1_LRST_SHIFT				5
+#define TI816X_USB1_LRST_MASK				BITFIELD(5, 5)
+
+/* Used by RM_DEFAULT_RSTCTRL */
+#define TI816X_USB2_LRST_SHIFT				6
+#define TI816X_USB2_LRST_MASK				BITFIELD(6, 6)
+
+/* Used by RM_DEFAULT_RSTCTRL */
+#define TI816X_PCI_LRST_SHIFT				7
+#define TI816X_PCI_LRST_MASK				BITFIELD(7, 7)
+
+#endif
diff --git a/arch/arm/mach-omap2/prm.h b/arch/arm/mach-omap2/prm.h
index 8ace464..9f199d4 100644
--- a/arch/arm/mach-omap2/prm.h
+++ b/arch/arm/mach-omap2/prm.h
@@ -26,6 +26,8 @@
 		OMAP2_L4_IO_ADDRESS(OMAP4430_PRM_BASE + (module) + (reg))
 #define OMAP44XX_CHIRONSS_REGADDR(module, reg)				\
 		OMAP2_L4_IO_ADDRESS(OMAP4430_CHIRONSS_BASE + (module) + (reg))
+#define TI816X_PRM_REGADDR(module, reg)					\
+		TI816X_L4_SLOW_IO_ADDRESS(TI816X_PRCM_BASE + (module) + (reg))
 
 #include "prm44xx.h"
 
@@ -227,6 +229,12 @@
 #define OMAP4_PM_PWSTCTRL				0x0000
 #define OMAP4_PM_PWSTST					0x0004
 
+/* TI816X specific registers */
+#define TI816X_PRM_DEVICE_RSTCTRL			0x00A0
+#define TI816X_PM_DEFAULT_PWRSTCTRL			0x0000
+#define TI816X_PM_DEFAULT_PWRSTST			0x0004
+#define TI816X_RM_DEFAULT_RSTCTRL			0x0010
+#define TI816X_RM_DEFAULT_RSTST				0x0014
 
 #ifndef __ASSEMBLER__
 
@@ -303,6 +311,8 @@ int omap2_prm_deassert_hardreset(s16 prm_mod, u8 shift);
 #define OMAP_RST_DPLL3					(1 << 2)
 #define OMAP_RST_GS					(1 << 1)
 
+/* TI816X specific bits for PRM_DEVICE module */
+#define TI816X_GLOBAL_RST_COLD				BIT(1)
 
 /*
  * Bits common to module-shared registers
-- 
1.7.0.2

